src/lambda/cached.rs

```diff
@@ -0,0 +1,97 @@
+use core::future::Future;
+use std::num::NonZeroU64;
+use std::pin::Pin;
+use std::sync::{Arc, RwLock};
+
+use anyhow::Result;
+use async_graphql_value::ConstValue;
+
+use super::{Concurrent, Eval, EvaluationContext, Expression, ResolverContextLike, IO};
+use crate::lambda::has_io::HasIO;
+
+pub trait CacheKey<Ctx> {
+    fn cache_key(&self, ctx: &Ctx) -> u64;
+}
+
+#[derive(Clone, Debug)]
+pub enum Cached {
+    IOCache(IOCache),
+    NonIOCache(NonIOCache),
+}
+
+#[derive(Clone, Debug)]
+pub struct IOCache {
+    max_age: NonZeroU64,
+    expr: Box<Expression>,
```

Comment:
```suggestion
    expr: Box<Expression>,
    cache_key: Fn(Ctx) -> u64
```

Can we inject the cache key function directly into the IOCache? That way we won't need to pattern match and look further into what is the actual expression and get it's key.